File: k_polyakov/4843.py
def spec(x):
    if x < 0:
        n = -x
        s = 0
        while n > 0:
            s += n%3
            n //= 3
        return s == 12
    return False

with open('27-91a.txt') as f:
    N, K, D = map(int, f.readline().split())

    INF = 10**30
    T = [[INF]*D for i in range(K)]
    
    special, s, m = 0, 0, 0

    for j in range(N):
        x = int(f.readline())
        s += x
        l = j + 1
        special += spec(x)
        if special%K == 0 and l%D == 0:
            m = max(m, s)
        if T[special%K][l%D] != INF:
            m = max(m, s - T[special%K][l%D])

        T[special%K][l%D] = min(T[special%K][l%D], s)

    print(m)




        



   




        
    

File: k_polyakov/4843.cpp

#include <iostream>
using namespace std;
#include <fstream>
#include <vector>


bool unusual(int x) {
	if (x > 0) { return false; }
	else {
		int s = 0;
		int t = -x;
		while (t > 0) { s += t % 3; t /= 3; }

		if (s == 12) { return true; }
		else return false;
	}
}



int main()
{
	fstream f;
	f.open("C:/Users/lenya/Desktop/27-я задача/С++/27/27-90b.txt");
	int m, total_s, N, x, u_count, K, D, total_len;
	total_s = 0; total_len = 0;
	u_count = 0; m = 0;
	f >> N >> K >> D;
	vector  < vector <int> > prefix_s(K, vector<int>(D, 100000000));
	for (int i = 0; i < N; i++) {
		total_len++;
		f >> x;
		total_s += x;
		if (unusual(x) == true) { ++u_count; }
		if (((prefix_s[u_count % K][total_len % D] != 100000000))) {
			m = max(m, total_s - prefix_s[u_count%K][total_len%D]); }
		prefix_s[u_count % K][total_len % D] = min(prefix_s[u_count % K][total_len % D], total_s);
	}
	cout << m;
}

